### PR TITLE
Skip segment for `MatchNoDocsQuery` filters

### LINKED ISSUE #94637: Aggs and constant_keyword
### Description

It'd be lovely is a `filters` or `filter` agg that touches a `constant_keyword` would properly rewrite itself to `match_all` and `match_none` and "go fast", at least when it's at the root of the aggs tree. Theoretically it should "just work" but it doesn't seem to. Also! The `terms` agg should be able to act on there being only a single possible value. Maybe it can even act on the `match_all`/`match_none` nature of the `constant_keyword`. I'm not sure how, but it does feel like there are optimizations in there. Particularly when `terms` is the top level agg.

### PR DESCRIPTION
**Pull request by kkrik-es**
When a query of a filter gets rewritten to `MatchNoDocsQuery`, segments will not produce any results. We can therefore skip processing such segments.

This applies to `FilterByFilterAggregator` and `FiltersAggregator`, as well as to `TermsAggregator` when it uses `StringTermsAggregatorFromFilters` internally; the latter is an adapter aggregator wrapping `FilterByFilterAggregator`.

Fixes #94637

### CODE REVIEW

**Comment by kkrik-es** [comment]
@elasticsearchmachine run elasticsearch-ci/part-2

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/FiltersAggregator.java [hunk 1]
```diff
@@ -289,6 +290,10 @@ static class Compatible extends FiltersAggregator {
 
         @Override
         protected LeafBucketCollector getLeafCollector(AggregationExecutionContext aggCtx, LeafBucketCollector sub) throws IOException {
+            if (filters().isEmpty() || (filters().size() == 1 && filters().get(0).query() instanceof MatchNoDocsQuery)) {
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/FiltersAggregator.java:293]: Maybe we should check whether all filters are `MatchNoDocsQuery`?

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/FilterByFilterAggregator.java [hunk 2]
```diff
@@ -233,7 +234,7 @@ private FilterByFilterAggregator(
     @Override
     protected LeafBucketCollector getLeafCollector(AggregationExecutionContext aggCtx, LeafBucketCollector sub) throws IOException {
         assert scoreMode().needsScores() == false;
-        if (filters().size() == 0) {
+        if (filters().isEmpty() || (filters().size() == 1 && filters().get(0).query() instanceof MatchNoDocsQuery)) {
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/FilterByFilterAggregator.java:237]: Check whether all filters are match no docs?

**Review by martijnvg — COMMENTED** [review]
Looks good, left a question.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/FiltersAggregator.java (hunk 1 above)
Comment (kkrik-es) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/FiltersAggregator.java:293]: Wouldn't these get collapsed into a single MatchNoDocsQuery during merging/rewriting?

Comment (kkrik-es) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/FiltersAggregator.java]: Updated to play it safe.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/FilterByFilterAggregator.java (hunk 2 above)
Comment (kkrik-es) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/FilterByFilterAggregator.java]: Done.

**Review by martijnvg — APPROVED** [review]
lgtm

**Comment by kkrik-es** [comment]
@elasticsearchmachine run elasticsearch-ci/part-3

**Comment by kkrik-es** [comment]
The optimization doesn't apply when `other` bucket is added to the filter spec. Good thing that a test caught this, added a yaml test too.

**Pull request merged**